eval/src/vespa/eval/tensor/sparse/sparse_tensor_t.cpp

```diff
@@ -0,0 +1,251 @@
+
+template<typename T>
+SparseTensorT<T>::SparseTensorT(eval::ValueType type_in, SparseTensorIndex index_in, std::vector<T> values_in)
+    : SparseTensor(std::move(type_in), std::move(index_in)),
+      _values(std::move(values_in))
+{
+}
+
+template<typename T>
+SparseTensorT<T>::~SparseTensorT() = default;
+
+template<typename T>
+TypedCells
+SparseTensorT<T>::cells() const
+{
+    return TypedCells(_values);
+}
+
+template<typename T>
+double
+SparseTensorT<T>::as_double() const
+{
+    double result = 0.0;
+    for (double v : _values) {
+        result += v;
+    }
+    return result;
+}
+
+template<typename T>
+void
+SparseTensorT<T>::accept(TensorVisitor &visitor) const
+{
+    TensorAddressBuilder addrBuilder;
+    TensorAddress addr;
+    for (const auto & kv : index().get_map()) {
+        SparseTensorAddressDecoder decoder(kv.first);
+        addrBuilder.clear();
+        for (const auto &dimension : fast_type().dimensions()) {
+            auto label = decoder.decodeLabel();
+            if (label.size() != 0u) {
+                addrBuilder.add(dimension.name, label);
+            }
+        }
+        assert(!decoder.valid());
+        addr = addrBuilder.build();
+        visitor.visit(addr, get_value(kv.second));
+    }
+}
+
+template<typename T>
+std::unique_ptr<Tensor>
+SparseTensorT<T>::add(const Tensor &arg) const
+{
+    const SparseTensor *rhs = dynamic_cast<const SparseTensor *>(&arg);
+    if (!rhs) {
+        return Tensor::UP();
+    }
+    SparseTensorAdd<T> adder(fast_type(), index(), _values);
+    rhs->accept(adder);
+    return adder.build();
+}
+
+template<typename T>
+Tensor::UP
+SparseTensorT<T>::apply(const CellFunction &func) const
+{
+    std::vector<T> new_values;
+    new_values.reserve(_values.size());
+    for (T v : _values) {
+        new_values.push_back(func.apply(v));
+    }
+    return std::make_unique<SparseTensorT<T>>(fast_type(), index(), std::move(new_values));
+}
+
+template<typename T>
+Tensor::UP
+SparseTensorT<T>::clone() const
+{
+    return std::make_unique<SparseTensorT<T>>(fast_type(), index(), _values);
+}
+
+template<typename T>
+Tensor::UP
+SparseTensorT<T>::join(join_fun_t function, const Tensor &arg) const
+{
+    const SparseTensor *rhs = dynamic_cast<const SparseTensor *>(&arg);
+    if (!rhs) {
+        return Tensor::UP();
+    }
+    const auto & lhs_type = fast_type();
+    const auto & rhs_type = rhs->fast_type();
+    auto res_type = eval::ValueType::join(lhs_type, rhs_type);
+    if (function == eval::operation::Mul::f) {
+        if (lhs_type.dimensions() == rhs_type.dimensions()) {
+            return typify_invoke<1,eval::TypifyCellType,FastSparseJoin<T>>(
+                    rhs_type.cell_type(),
+                    *this, *rhs, std::move(res_type));
+        }
+    }
+    return typify_invoke<2,eval::TypifyCellType,GenericSparseJoin<T>>(
+            rhs_type.cell_type(), res_type.cell_type(),
+            *this, *rhs, std::move(res_type), function);
```

Comment:
While this likely works due to `res_type._cell_type` not being cleared due to the move constructor being called, it would likely be better to adjust the code to ensure that `res_type.cell_type()` is evaluated first.

Reply:
evaluating std::move(value) will not change the value, since it 'only' casts the value to an rvalue. I thought that all parameters were evaluated (in undefined order) before the function was actually called, and that the copying of parameters into the local scope of the function was part of evaluating the function itself. I might be wrong, but changing the code will make it clearer and safer any way, so I am good with that. Maybe @vekterli can answer what the standard says about this?

Reply:
Had a quick rummage through the C++20 ISO draft.

**§ 6.9.1.10**

> Except where noted, evaluations of operands of individual operators and of subexpressions of individual expressions are unsequenced. (...) If a side effect on a memory location (6.7.1) is unsequenced relative to either another side effect on the same memory location or a value computation using the value of any object in the same memory location, and they are not potentially concurrent (6.9.2), the behavior is undefined.

**§ 6.9.1.11**
> When calling a function (whether or not the function is inline), every value computation and side effect
associated with any argument expression, or with the postfix expression designating the called function, is
sequenced before execution of every expression or statement in the body of the called function.

**§ 7.6.1.2.7** (emphasis mine)

> The initialization and destruction of each parameter occurs within the context of the _calling_ function. 

Unless there are some more sneaky sequencing constraints I'm not aware of, this implies to me that the object move-construction from the rvalue (and subsequent invalidation of the object pointed to by the rvalue) _may_ take place before it's used as part of another parameter.